nuxeo-quota-automation/src/main/java/org/nuxeo/ecm/quota/automation/RecomputeQuotaStatistics.java

```diff
@@ -0,0 +1,106 @@
+import org.nuxeo.ecm.automation.core.annotations.Operation;
+import org.nuxeo.ecm.automation.core.annotations.OperationMethod;
+import org.nuxeo.ecm.automation.core.annotations.Param;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.platform.userworkspace.api.UserWorkspaceService;
+import org.nuxeo.ecm.quota.QuotaStatsService;
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * Recomputes Quota size statistics.
+ *
+ * @since 10.1
+ */
+@Operation(id = RecomputeQuotaStatistics.ID, category = "Quotas", //
+        label = "Recompute quota statistics on documents, optionally only for a tenant, user or path")
+public class RecomputeQuotaStatistics {
+
+    public static final String ID = "Quotas.RecomputeStatistics";
+
+    public static final String SIZE_UPDATER = "documentsSizeUpdater";
+
+    @Param(name = "tenantId", required = false)
+    protected String tenantId;
+
+    @Param(name = "username", required = false)
+    protected String username;
+
+    @Param(name = "path", required = false)
+    protected String path;
+
+    @Param(name = "updaterName", required = false)
+    protected String updaterName;
+
+    @Context
+    protected CoreSession session;
+
+    @Context
+    protected QuotaStatsService quotaStatsService;
+
+    @OperationMethod
+    public String run() throws OperationException {
+        String docPath;
+        if (tenantId != null) {
+            if (username != null || path != null) {
+                throw newException();
+            }
+            docPath = getTenantPath();
+        } else if (username != null) {
+            if (path != null) {
+                throw newException();
+            }
+            docPath = getUserPersonalWorkspacePath();
+        } else {
+            docPath = path; // may be null
+        }
+        if (updaterName == null) {
+            updaterName = SIZE_UPDATER;
+        }
+        String repositoryName = session.getRepositoryName();
+        quotaStatsService.launchInitialStatisticsComputation(updaterName, repositoryName, docPath);
+        return quotaStatsService.getProgressStatus(updaterName, repositoryName);
+    }
+
+    protected OperationException newException() {
+        return new OperationException("Only one of tenantId, username or path can be defined");
+    }
+
+    protected String getTenantPath() throws OperationException {
+        // TODO this should use the multi-tenant service instead of assuming that tenants are per-domain
+        if (tenantId.contains("/")) {
+            throw new OperationException("Invalid tenantId: " + tenantId);
+        }
+        return "/" + tenantId;
+    }
+
+    protected String getUserPersonalWorkspacePath() {
+        UserWorkspaceService uws = Framework.getService(UserWorkspaceService.class);
+        DocumentModel userWorkspace = uws.getUserPersonalWorkspace(username, session.getRootDocument());
+        return userWorkspace.getPathAsString();
```

Comment:
Could `userWorkspace` be `null`? eg. if an invalid username is provided.

Reply:
I'll add a check.